src/hip_memory.cpp

```diff
@@ -320,20 +320,34 @@ void generic_copy(void* __restrict dst, const void* __restrict src, size_t n,
 inline
 void memcpy_impl(void* __restrict dst, const void* __restrict src, size_t n,
                  hipMemcpyKind k) {
+    auto si{info(src)};
+    auto di{info(dst)};
+
+    if (!is_large_BAR){
+       // Pointer info takes presidence over hipMemcpyKind
+       // if there is mismatch b/w Memcpy kind and dst/src pointer
+       // E.g. dst(host pointer),src(device pointer) and hipMemcpyKind set as hipMemcpyHostToDevice
+       if (di.size == is_cpu_owned && si.size == is_cpu_owned)
```

Comment:
Shouldn't these checks be lowered into the `h2d_copy`, `d2h_copy` and `generic_copy`'s below?  We are introducing extra overhead for no reason by running `info` and doing the checks here. 

For instance, once we've committed to a `hipMemcpyHostToDevice`, we are already determining the source / destination pointer information (e.g., on line 267).  Why not put the statement:

```
si.size == is_cpu_owned && di.size != is_cpu_owned
```

 in `h2d_copy`, and return a `hipInvalidValue` or some other error code at that point if it fails?

Reply:
For API hipMemcpyHostToDevice & hipMemcpyDeviceToHost it is fine but for H2H case we need to have extra info()s anyway. So if you see only one extra info() that is being added here.